src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs

```diff
@@ -143,43 +141,52 @@ private void Connect(Socket socket, EndPoint endPoint, CancellationToken cancell
                     {
                         socket.Connect(endPoint);
                     }
-                    connected = true;
+                    ChangeState(2);
                     return;
                 }
                 catch
                 {
-                    if (!cancelled && !timedOut)
+                    switch (state)
```

Comment:
These changes are not required to fix the race condition but are an attempt to satisfy your concerns so that you can approve the pull request.